applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp

```diff
@@ -23,25 +25,163 @@
+
+            } else if constexpr(std::is_same_v<TDataType, array_1d<double, 3>>) {
+                AtomicAddVector( r_node.FastGetSolutionStepValue(rNodalVariable), (elem_rVariable) );
+                node_element_count[r_node.Id()] += 1;
+            } else {
+                static_assert(!std::is_same_v<TDataType, TDataType>, "Unsupported data type.");
+            }
+        }
+    });
+
+    block_for_each(rModelPart.Nodes(), [&](Node& rNode){
+        if (node_element_count.find(rNode.Id()) != node_element_count.end()){
+            rNode.FastGetSolutionStepValue(rNodalVariable) /= node_element_count[rNode.Id()];
+        }
+        else{
+            KRATOS_ERROR << "Node " << rNode.Id() << " has no associated elements." << std::endl;
         }
     });
 
     rModelPart.GetCommunicator().AssembleCurrentData(rNodalVariable);
 }
 
+template<class TDataType>
+void ConversionUtilities::ConvertNodalDataToElementalDataDirect(
+    ModelPart& rModelPart,
+    const Variable<TDataType>& rElementalVariable,
+    const Variable<TDataType>& rNodalVariable )
+{
+    // prepare elemental variable
+    VariableUtils().SetNonHistoricalVariableToZero(rElementalVariable, rModelPart.Elements());
+
+    block_for_each(rModelPart.Elements(), [&](Element& rElement){
+        const std::size_t num_nodes = rElement.GetGeometry().PointsNumber();
+
+        if constexpr(std::is_same_v<TDataType, double>) {
+                double temp = 0.0;
+                for (auto& r_node : rElement.GetGeometry().Points()){
+                    temp += r_node.FastGetSolutionStepValue(rNodalVariable) / num_nodes;
+                } 
+                rElement.SetValue(rElementalVariable, temp);
+            }
+        else if constexpr(std::is_same_v<TDataType, array_1d<double, 3>>) {
+            array_1d<double, 3> temp = ZeroVector(3);
+            for (auto& r_node : rElement.GetGeometry().Points()){
+                temp += r_node.FastGetSolutionStepValue(rNodalVariable) / num_nodes;
+            }
+            rElement.SetValue(rElementalVariable, temp);
+        }
+        
+        else {
+                static_assert(!std::is_same_v<TDataType, TDataType>, "Unsupported data type.");
+            }
+    });
+
+    rModelPart.GetCommunicator().AssembleCurrentData(rElementalVariable);
+}
+
+template<class TDataType>
+void ConversionUtilities::ConvertNodalDataToElementalDataTranspose(
+    ModelPart& rModelPart,
+    const Variable<TDataType>& rElementalVariable,
+    const Variable<TDataType>& rNodalVariable )
+{
+    // prepare elemental variable
+    VariableUtils().SetNonHistoricalVariableToZero(rElementalVariable, rModelPart.Elements());
+
+    // count number of shared elements for each node
+    std::map<int, int> node_element_count;
+    block_for_each(rModelPart.Elements(), [&](Element& rElement){
+        for (auto& r_node : rElement.GetGeometry().Points()){
+            node_element_count[r_node.Id()] += 1;
+        }
+    });
```

Comment:
This is a gigantic race condition. You're inserting items into the container in parallel **and** incrementing the mapped entries in parallel without any protections too.

Using an ordered map is also unnecessary here.

Reply:
I have changed to AtomicAdd.

Reply:
> I have changed to AtomicAdd.

That's not enough because you're still inserting entries into the map in parallel. You can either initialize all entries first in serial, or construct a local map in each thread and merge them in the end.

Reply:
> Using an ordered map is also unnecessary here.